File: 02.AddTwoNumbers/function.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val  int
 *     Next *ListNode
 * }
 */
func addTwoNumbers(l1 *ListNode, l2 *ListNode) *ListNode {
    dummy := &ListNode{Val: 0}
    current := dummy
    carry := 0

    for l1 != nil || l2 != nil || carry > 0 {
        x := 0
        if l1 != nil {
            x = l1.Val
        }
        y := 0
        if l2 != nil {
            y = l2.Val
        }

        sum := x + y + carry
        carry = sum / 10

        current.Next = &ListNode{Val: sum % 10}
        current = current.Next

        if l1 != nil {
            l1 = l1.Next
        }
        if l2 != nil {
            l2 = l2.Next
        }
    }
    return dummy.Next
}

File: 02.AddTwoNumbers/function.rb
# Definition for singly-linked list.
# class ListNode
#     attr_accessor :val, :next
#     def initialize(val = 0, _next = nil)
#         @val = val
#         @next = _next
#     end
# end
def add_two_numbers(l1, l2)
    dummy = ListNode.new(0)
    current = dummy
    carry = 0

    while l1 || l2 || carry > 0
        x = l1 ? l1.val : 0
        y = l2 ? l2.val : 0
        sum = x + y + carry

        carry = sum / 10
        current.next = ListNode.new(sum % 10)
        current = current.next

        l1 = l1.next if l1
        l2 = l2.next if l2
    end

    dummy.next
end
